File: test/src/test/java/org/camunda/bpm/extension/process_test_coverage/rules/ClassCoverageSystemPropertyTest.java
package org.camunda.bpm.extension.process_test_coverage.rules;

import org.camunda.bpm.engine.test.Deployment;
import org.camunda.bpm.extension.process_test_coverage.junit.rules.TestCoverageProcessEngineRule;
import org.camunda.bpm.extension.process_test_coverage.junit.rules.TestCoverageProcessEngineRuleBuilder;
import org.junit.AfterClass;
import org.junit.ClassRule;
import org.junit.Rule;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

import static org.hamcrest.Matchers.*;

/**
 * Test case starting an in-memory database-backed Process Engine.<br>
 * In your tests don't set the property using <code>System.setProperty</code> -
 * use your runtime environment to do that.
 */
public class ClassCoverageSystemPropertyTest {

    static final double EXPECTED = CoverageTestProcessConstants.PATH_B_ELEMENTS.length;
    static final double ALL = CoverageTestProcessConstants.ALL_ELEMENTS.length;
    static final double EXPECTED_COVERAGE = EXPECTED / ALL;

    static {
        setSysProperty();
    }

    public static void setSysProperty() {
        System.setProperty(TestCoverageProcessEngineRuleBuilder.DEFAULT_ASSERT_AT_LEAST_PROPERTY,
                "" + EXPECTED_COVERAGE);
    }

    @AfterClass
    public static void delSysProperty() {
        System.clearProperty(TestCoverageProcessEngineRuleBuilder.DEFAULT_ASSERT_AT_LEAST_PROPERTY);
    }

    @ClassRule
    @Rule
    public static TestCoverageProcessEngineRule classRule = 
            TestCoverageProcessEngineRuleBuilder.create()
            	.optionalAssertCoverageAtLeastProperty(TestCoverageProcessEngineRuleBuilder.DEFAULT_ASSERT_AT_LEAST_PROPERTY)
            	.build();

    @Test
    @Deployment(resources = CoverageTestProcessConstants.BPMN_PATH)
    public void testPathB() {
        Map<String, Object> variables = new HashMap<String, Object>();
        variables.put("path", "B");
        classRule.getRuntimeService().startProcessInstanceByKey(CoverageTestProcessConstants.PROCESS_DEFINITION_KEY, variables);

        classRule.addTestMethodCoverageAssertionMatcher("testPathB", lessThan(EXPECTED_COVERAGE + 0.0001));
    }

}

File: extension/junit5-platform-8/src/test/kotlin/org/camunda/bpm/extension/process_test_coverage/junit5/ClassCoverageSystemPropertyTest.kt
package org.camunda.bpm.extension.process_test_coverage.junit5

import io.camunda.zeebe.client.ZeebeClient
import io.camunda.zeebe.process.test.api.ZeebeTestEngine
import io.camunda.zeebe.process.test.extension.testcontainer.ZeebeProcessTest
import org.assertj.core.api.HamcrestCondition
import org.camunda.bpm.extension.process_test_coverage.junit5.CoverageTestProcessConstants.deploy
import org.camunda.bpm.extension.process_test_coverage.junit5.ProcessEngineCoverageExtension.Builder.Companion.DEFAULT_ASSERT_AT_LEAST_PROPERTY
import org.hamcrest.Matchers
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.RegisterExtension
import java.time.Duration


@ZeebeProcessTest
class ClassCoverageSystemPropertyTest {

    companion object {

        val EXPECTED: Double = CoverageTestProcessConstants.PATH_B_ELEMENTS.size.toDouble()
        val ALL: Double = CoverageTestProcessConstants.ALL_ELEMENTS.size.toDouble()
        val EXPECTED_COVERAGE = EXPECTED / ALL

        init {
            System.setProperty(DEFAULT_ASSERT_AT_LEAST_PROPERTY, "$EXPECTED_COVERAGE")
        }

        @AfterAll
        @JvmStatic
        fun delSysProperty() {
            System.clearProperty(DEFAULT_ASSERT_AT_LEAST_PROPERTY)
        }

        @RegisterExtension
        @JvmField
        var extension: ProcessEngineCoverageExtension = ProcessEngineCoverageExtension.builder()
                .optionalAssertCoverageAtLeastProperty(DEFAULT_ASSERT_AT_LEAST_PROPERTY)
                .build()
    }

    private lateinit var client: ZeebeClient
    private lateinit var engine: ZeebeTestEngine

    @Test
    fun testPathB() {
        deploy(client)
        val variables: MutableMap<String, Any> = HashMap()
        variables["path"] = "B"
        client.newCreateInstanceCommand().bpmnProcessId(CoverageTestProcessConstants.PROCESS_DEFINITION_KEY).latestVersion().variables(variables).send().join()
        extension.addTestMethodCoverageCondition("testPathB", HamcrestCondition(Matchers.lessThan(EXPECTED_COVERAGE + 0.0001)))
        engine.waitForIdleState(Duration.ofSeconds(5))
    }


}